apps/emqx_durable_storage/src/emqx_ds_snapshot_manager.erl

```diff
@@ -0,0 +1,428 @@
+%% See the License for the specific language governing permissions and
+%% limitations under the License.
+%%--------------------------------------------------------------------
+-module(emqx_ds_snapshot_manager).
+
+-behaviour(gen_server).
+
+%% API
+-export([
+    start_link/2,
+
+    begin_read/2,
+    read_chunk/3,
+
+    begin_accept/2,
+    accept_chunk/2,
+    complete_accept/2
+]).
+
+%% `gen_server' API
+-export([
+    init/1,
+    terminate/2,
+
+    handle_call/3,
+    handle_cast/2,
+    handle_info/2
+]).
+
+-include_lib("snabbkaffe/include/trace.hrl").
+
+%%------------------------------------------------------------------------------
+%% Type declarations
+%%------------------------------------------------------------------------------
+
+-define(via(SHARD_ID), {via, gproc, {n, l, {?MODULE, SHARD_ID}}}).
+
+-type shard_id() :: {emqx_ds:db(), emqx_ds_replication_layer:shard_id()}.
+
+-type state() :: #{
+    rs := rs() | undefined,
+    ws := ws() | undefined
+}.
+
+-type ra_state() :: emqx_ds_replication_layer:ra_state().
+
+%% Read state.
+-record(rs, {
+    phase :: machine_state | reader_caller_proc | storage_snapshot,
+    started_at :: _Time :: integer(),
+    state :: ra_state() | undefined,
+    reader_caller_proc :: {pid(), reference()} | undefined,
+    reader :: emqx_ds_storage_snapshot:reader() | undefined
+}).
+
+%% Write state.
+-record(ws, {
+    phase :: machine_state | reader_caller_proc | storage_snapshot,
+    started_at :: _Time :: integer(),
+    dir :: file:filename(),
+    meta :: ra_snapshot:meta(),
+    state :: ra_state() | undefined,
+    reader_caller_proc :: {pid(), reference()} | undefined,
+    writer :: emqx_ds_storage_snapshot:writer() | undefined
+}).
+
+-type rs() :: #rs{}.
+-type ws() :: #ws{}.
+
+-type ext_rs() :: #{shard_id := shard_id()}.
+-type ext_ws() :: #{phase := machine_state, ws := ws()} | #{shard_id := shard_id()}.
+
+%% call/cast/info events
+-record(start_reader, {rs :: rs(), caller :: pid()}).
+-record(read_chunk, {size :: non_neg_integer()}).
+-record(begin_accept, {ws :: ws()}).
+-record(accept_chunk, {chunk :: binary()}).
+-record(complete_accept, {chunk :: binary()}).
+
+-export_type([ext_rs/0, ext_ws/0]).
+
+%%------------------------------------------------------------------------------
+%% API
+%%------------------------------------------------------------------------------
+
+-spec start_link(emqx_ds:db(), emqx_ds_replication_layer:shard_id()) -> {ok, pid()}.
+start_link(DB, ShardId) ->
+    gen_server:start_link(?via({DB, ShardId}), ?MODULE, #{db => DB, shard_id => ShardId}, []).
+
+-spec begin_read(_SnapshotDir :: file:filename(), _Context :: #{}) ->
+    {ok, ra_snapshot:meta(), ext_rs()} | {error, _Reason :: term()}.
+begin_read(Dir, _Context) ->
+    RS0 = #rs{
+        phase = machine_state,
+        started_at = erlang:monotonic_time(millisecond)
+    },
+    case ra_log_snapshot:recover(Dir) of
```

Comment:
Now there's replication-level concerns trickled on that level for some reason. Should be much clearer if anything related to `ra_snapshot` behaviour implementation stays in the respective module.